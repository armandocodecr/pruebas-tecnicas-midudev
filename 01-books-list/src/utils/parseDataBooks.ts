import { IBookInterface } from "../interfaces/IBookInterfaces";

export function parseDataBooks(data: IBookInterface[], localStorageData?: IBookInterface[]): Array<IBookInterface> {
    return data.map((books) => {
      const { book } = books;
      if (!localStorageData) return books;
  
      return {
        book: { ...book },
        isFavorite: localStorageData
          ? localStorageData.some((currBook) => currBook.book.ISBN === books.book.ISBN)
          : false,
      };
    });
  }