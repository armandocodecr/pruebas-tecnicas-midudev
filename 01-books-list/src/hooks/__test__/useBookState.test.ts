import { test } from "vitest";
import { renderHook } from "@testing-library/react-hooks";
import { useMemo } from "react";
import dataBooks from "../../mook/books.json";
import { IBookInterface } from "../../interfaces/IBookInterfaces";
import { parseDataBooks } from "../../utils/parseDataBooks";
import { mookDataLocalStorage, mookDefaultData } from "./mook/mookDataBooks";

describe("Test of filteredBooks and parseBooks", () => {
  test("it should parse books based on local storage data", () => {
    const { state } = mookDataLocalStorage;
    const dataBooks: IBookInterface[] = parseDataBooks(
      mookDefaultData,
      state.favoriteBookStores
    );
    const dataBooksWithoutFavoriteBooks = parseDataBooks(mookDefaultData);

    expect(dataBooksWithoutFavoriteBooks).toEqual(mookDefaultData);
    //üëÜ Verifies that if we do not pass it a persistent state of the favorite books, it will return the original data.
    expect(dataBooks).toEqual([
      {
        book: {
          title: "El Se√±or de los Anillos",
          pages: 1200,
          genre: "Fantas√≠a",
          cover:
            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1566425108i/33.jpg",
          synopsis:
            "Una aventura √©pica en un mundo de fantas√≠a llamado la Tierra Media.",
          year: 1954,
          ISBN: "978-0618640157",
          author: {
            name: "J.R.R. Tolkien",
            otherBooks: ["El Hobbit", "El Silmarillion"],
          },
        },
        isFavorite: true,
      },
      {
        book: {
          title: "Juego de Tronos",
          pages: 694,
          genre: "Fantas√≠a",
          cover:
            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1273763400i/8189620.jpg",
          synopsis:
            "En un reino donde las estaciones duran a√±os, una batalla √©pica por el trono se desarrolla.",
          year: 1996,
          ISBN: "978-0553103540",
          author: {
            name: "George R. R. Martin",
            otherBooks: [
              "Choque de Reyes",
              "Tormenta de Espadas",
              "Fest√≠n de Cuervos",
            ],
          },
        },
        isFavorite: false,
      },
    ]);
  });

  test("should return the filter books by genre correctly", () => {
    const selectGenreValue = "Zombies";
    const books = parseDataBooks(dataBooks);
    const { result } = renderHook(() =>
      useMemo(() => {
        const booksFilter = selectGenreValue
          ? books.filter((book) =>
              book.book.genre
                .toLowerCase()
                .startsWith(selectGenreValue.toLowerCase())
            )
          : books;
        if (booksFilter.length === 0) return books;
        return booksFilter;
      }, [books, selectGenreValue])
    );

    expect(result.current).toMatchSnapshot([
      {
        book: {
          title: "Apocalipsis Zombie",
          pages: 444,
          genre: "Zombies",
          cover:
            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1422626176i/24762432.jpg",
          synopsis:
            "Un gallego se queda en casa en pleno apocalipsis zombie y acaba casi salvando el mundo",
          year: 2001,
          ISBN: "978-4444532611",
          author: {
            name: "Manel Loreiro",
            otherBooks: [],
          },
        },
      },
    ]);
  });

  test("should return the filter books by name correctly", () => {
    const inputValue = "El".toLowerCase();
    const books = parseDataBooks(dataBooks);
    const newArray: IBookInterface[] = books.filter((books) =>
      books.book.title.toLowerCase().startsWith(inputValue)
    );

    expect(newArray).toEqual([
      {
        book: {
          title: "El Se√±or de los Anillos",
          pages: 1200,
          genre: "Fantas√≠a",
          cover:
            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1566425108i/33.jpg",
          synopsis:
            "Una aventura √©pica en un mundo de fantas√≠a llamado la Tierra Media.",
          year: 1954,
          ISBN: "978-0618640157",
          author: {
            name: "J.R.R. Tolkien",
            otherBooks: ["El Hobbit", "El Silmarillion"],
          },
        }
      },
      {
        book: {
          title: "El resplandor",
          pages: 688,
          genre: "Terror",
          cover:
            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1641398308i/60038757.jpg",
          synopsis:
            "Una familia se muda a un hotel aislado para el invierno donde una presencia siniestra influye en el padre hacia la violencia.",
          year: 1977,
          ISBN: "978-0307743657",
          author: {
            name: "Stephen King",
            otherBooks: ["Carrie", "It"],
          },
        }
      },
    ]);
  });
});
