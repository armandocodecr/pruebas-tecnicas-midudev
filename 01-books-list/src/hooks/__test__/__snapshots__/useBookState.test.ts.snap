// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test of filteredBooks > should return the filter books correctly 1`] = `
[
  {
    "book": {
      "ISBN": "978-4444532611",
      "author": {
        "name": "Manel Loreiro",
        "otherBooks": [],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1422626176i/24762432.jpg",
      "genre": "Zombies",
      "pages": 444,
      "synopsis": "Un gallego se queda en casa en pleno apocalipsis zombie y acaba casi salvando el mundo",
      "title": "Apocalipsis Zombie",
      "year": 2001,
    },
  },
]
`;

exports[`Test of filteredBooks and parseBooks > it should parse books based on local storage data 1`] = `
[
  {
    "book": {
      "ISBN": "978-0618640157",
      "author": {
        "name": "J.R.R. Tolkien",
        "otherBooks": [
          "El Hobbit",
          "El Silmarillion",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1566425108i/33.jpg",
      "genre": "Fantasía",
      "pages": 1200,
      "synopsis": "Una aventura épica en un mundo de fantasía llamado la Tierra Media.",
      "title": "El Señor de los Anillos",
      "year": 1954,
    },
    "isFavorite": true,
  },
  {
    "book": {
      "ISBN": "978-0553103540",
      "author": {
        "name": "George R. R. Martin",
        "otherBooks": [
          "Choque de Reyes",
          "Tormenta de Espadas",
          "Festín de Cuervos",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1273763400i/8189620.jpg",
      "genre": "Fantasía",
      "pages": 694,
      "synopsis": "En un reino donde las estaciones duran años, una batalla épica por el trono se desarrolla.",
      "title": "Juego de Tronos",
      "year": 1996,
    },
    "isFavorite": false,
  },
  {
    "book": {
      "ISBN": "978-0747532699",
      "author": {
        "name": "J.K. Rowling",
        "otherBooks": [
          "Harry Potter y la cámara secreta",
          "Harry Potter y el prisionero de Azkaban",
          "Harry Potter y el cáliz de fuego",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1550337333i/15868.jpg",
      "genre": "Fantasía",
      "pages": 223,
      "synopsis": "Un niño descubre que es un mago y comienza una aventura en una escuela de magia.",
      "title": "Harry Potter y la piedra filosofal",
      "year": 1997,
    },
    "isFavorite": true,
  },
]
`;

exports[`Test of filteredBooks and parseBooks > it should return the information of the parsed books correctly according to the information in the local storage 1`] = `
[
  {
    "book": {
      "ISBN": "978-0618640157",
      "author": {
        "name": "J.R.R. Tolkien",
        "otherBooks": [
          "El Hobbit",
          "El Silmarillion",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1566425108i/33.jpg",
      "genre": "Fantasía",
      "pages": 1200,
      "synopsis": "Una aventura épica en un mundo de fantasía llamado la Tierra Media.",
      "title": "El Señor de los Anillos",
      "year": 1954,
    },
    "isFavorite": true,
  },
  {
    "book": {
      "ISBN": "978-0553103540",
      "author": {
        "name": "George R. R. Martin",
        "otherBooks": [
          "Choque de Reyes",
          "Tormenta de Espadas",
          "Festín de Cuervos",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1273763400i/8189620.jpg",
      "genre": "Fantasía",
      "pages": 694,
      "synopsis": "En un reino donde las estaciones duran años, una batalla épica por el trono se desarrolla.",
      "title": "Juego de Tronos",
      "year": 1996,
    },
    "isFavorite": false,
  },
  {
    "book": {
      "ISBN": "978-0747532699",
      "author": {
        "name": "J.K. Rowling",
        "otherBooks": [
          "Harry Potter y la cámara secreta",
          "Harry Potter y el prisionero de Azkaban",
          "Harry Potter y el cáliz de fuego",
        ],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1550337333i/15868.jpg",
      "genre": "Fantasía",
      "pages": 223,
      "synopsis": "Un niño descubre que es un mago y comienza una aventura en una escuela de magia.",
      "title": "Harry Potter y la piedra filosofal",
      "year": 1997,
    },
    "isFavorite": true,
  },
]
`;

exports[`Test of filteredBooks and parseBooks > should return the filter books by genre correctly 1`] = `
[
  {
    "book": {
      "ISBN": "978-4444532611",
      "author": {
        "name": "Manel Loreiro",
        "otherBooks": [],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1422626176i/24762432.jpg",
      "genre": "Zombies",
      "pages": 444,
      "synopsis": "Un gallego se queda en casa en pleno apocalipsis zombie y acaba casi salvando el mundo",
      "title": "Apocalipsis Zombie",
      "year": 2001,
    },
  },
]
`;

exports[`Test of filteredBooks and parseBooks > should return the filter books by name correctly 1`] = `
[
  {
    "book": {
      "ISBN": "978-4444532611",
      "author": {
        "name": "Manel Loreiro",
        "otherBooks": [],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1422626176i/24762432.jpg",
      "genre": "Zombies",
      "pages": 444,
      "synopsis": "Un gallego se queda en casa en pleno apocalipsis zombie y acaba casi salvando el mundo",
      "title": "Apocalipsis Zombie",
      "year": 2001,
    },
  },
]
`;

exports[`Test of filteredBooks and parseBooks > should return the filter books correctly 1`] = `
[
  {
    "book": {
      "ISBN": "978-4444532611",
      "author": {
        "name": "Manel Loreiro",
        "otherBooks": [],
      },
      "cover": "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1422626176i/24762432.jpg",
      "genre": "Zombies",
      "pages": 444,
      "synopsis": "Un gallego se queda en casa en pleno apocalipsis zombie y acaba casi salvando el mundo",
      "title": "Apocalipsis Zombie",
      "year": 2001,
    },
  },
]
`;
